lisp/format_test.go

```diff
@@ -0,0 +1,185 @@
+package lisp
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFormatStringSimple(t *testing.T) {
+	env := NewEnv(nil)
+	format := String("Swap a symbol here {}")
+	arg1 := String("yeah")
+	args := &LVal{Cells: []*LVal{format, arg1}}
+	formatted := builtinFormatString(env, args)
+	assert.Equal(t, "Swap a symbol here yeah", formatted.Str)
+}
+
+func TestFormatStringComplex(t *testing.T) {
+	env := NewEnv(nil)
+	format := String("Swap a symbol here ")
+	built := format.Str
+	args := []*LVal{format}
+	for x := 0; x < 100; x++ {
+		args[0] = String(fmt.Sprintf("%s {}, blah blah,", args[0].Str))
+		built = fmt.Sprintf("%s %d, blah blah,", built, x)
+		args = append(args, String(strconv.Itoa(x)))
+		formatted := builtinFormatString(env, &LVal{Cells: args})
+		assert.Equal(t, built, formatted.Str)
+	}
+}
+
+func TestFormatStringUnmatched(t *testing.T) {
+	env := NewEnv(nil)
+	format := String("Swap a symbol here }")
+	arg1 := String("yeah")
+	args := &LVal{Cells: []*LVal{format, arg1}}
+	formatted := builtinFormatString(env, args)
+	assert.Equal(t, LError, formatted.Type)
+}
+
+func TestTokenizeFormatString(t *testing.T) {
+	x := tokenizeFormatString("This is a formatted {} string yeah {}.")
+	assert.Equal(t, []formatToken{
+		{typ: formatText, text: "This is a formatted "},
+		{typ: formatOpen, text: "{"},
+		{typ: formatClose, text: "}"},
+		{typ: formatText, text: " string yeah "},
+		{typ: formatOpen, text: "{"},
+		{typ: formatClose, text: "}"},
+		{typ: formatText, text: "."},
+	}, x)
+}
+
+var formatTokenRegexp = regexp.MustCompile(`\{(\d+)\}`)
+
+// var formatBadTokenPattern = regexp.MustCompile(`(?:\{[^\d\}]|[^\{\d]\})`) // This is slooooow...
+// var formatBadTokenPattern = regexp.MustCompile(`\{[^\d\}]`) // This is also slow enough that I'd ditch it
+
+func quickformatBuiltin(env *LEnv, args *LVal) *LVal {
+	format := args.Cells[0]
+	//if formatBadTokenPattern.MatchString(format.Str) {
+	//return env.Errorf("unmatched token")
```

Comment:
I think there should be a test/error for unmatched tokens.  The existing format-string does this.

Reply:
Ah. Maybe this was trying to catch unclosed curly braces? What I meant was that format-string will report an error if you don't supply enough values.

```
elps> (format-string "hello {}, it's {} to meet you" "world")
stdin:1: lisp:format-string: too many formatting direcives for supplied values
Stack Trace [1 frames -- entrypoint last]:
  height 0: stdin:1: lisp:format-string
```

Reply:
This should do that - it's just `fmt.Sprintf` inside, so it'll handle it the same way that Go does with a warning in the output.

Reply:
Right. fmt just outputs warning when there's a mismatch between argument and substitutions. But I don't think that is very good behavior for a language like lisp. I think a runtime error is preferable.

The annoying thing about the fmt package is that format argument mismatches can literally go unnoticed for onths or years. I think they chose to do that instead of forcing programmers to check errors for every formatting call. I've never really seen another argument for why formatting _should_ work that way (though I guess I haven't looked hard if you had one).

In dynamic languages like python and ruby format argument mismatches cause runtime exceptions. Even if it's not completely preferable, I think it's more expected of a language like elps.